File: slice.py
import argparse
import os
import sys

def read_slice(file_path, start, size):
    try:
        with open(file_path, 'rb') as f:
            f.seek(start)
            return f.read(size)
    except Exception as e:
        print(f"Error reading file: {e}", file=sys.stderr)
        sys.exit(1)

def trim_truncated_lines(data, start):
    lines = data.splitlines(keepends=True)

    # Remove first line if start > 0
    if start > 0 and lines:
        lines = lines[1:]

    # Remove last line if not ending in newline
    if lines and not lines[-1].endswith(b'\n'):
        lines = lines[:-1]

    return b''.join(lines)

def main():
    parser = argparse.ArgumentParser(description='Extract a byte slice from a file.')
    parser.add_argument('--start', type=int, required=True, help='Byte offset to start (0-based)')
    parser.add_argument('--size', type=int, required=True, help='Number of bytes to read')
    parser.add_argument('--file', required=True, help='File to extract from')
    parser.add_argument('--full-lines-only', action='store_true', help='Remove truncated lines at start/end')

    args = parser.parse_args()

    if args.start < 0:
        parser.error("--start must be >= 0")
    if args.size <= 0:
        parser.error("--size must be > 0")
    if not os.path.isfile(args.file):
        parser.error(f"File not found or not a regular file: {args.file}")

    try:
        data = read_slice(args.file, args.start, args.size)
        if args.full_lines_only:
            data = trim_truncated_lines(data, args.start)
        sys.stdout.buffer.write(data)
    except Exception as e:
        print(f"Unexpected error: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == '__main__':
    main()

File: slice.rb
require 'optparse'

options = {
  start: nil,
  size: nil,
  file: nil,
  full_lines_only: false
}

OptionParser.new do |opts|
  opts.banner = "Usage: slice.rb --start OFFSET --size BYTES --file FILENAME [--full-lines-only]"

  opts.on("--start OFFSET", Integer, "Byte offset to start reading") { |v| options[:start] = v }
  opts.on("--size BYTES", Integer, "Number of bytes to read") { |v| options[:size] = v }
  opts.on("--file FILENAME", String, "File to extract from") { |v| options[:file] = v }
  opts.on("--full-lines-only", "Remove truncated lines at start/end") { options[:full_lines_only] = true }
  opts.on("--help", "Prints this help") do
    puts opts
    exit
  end
end.parse!

if options[:file].nil? || !File.file?(options[:file])
  STDERR.puts "Error: File not found or invalid."
  exit 1
end

if options[:start].nil? || options[:size].nil? || options[:start] < 0 || options[:size] <= 0
  STDERR.puts "Error: --start must be >= 0 and --size must be > 0."
  exit 1
end

def read_slice(filename, start, size)
  File.open(filename, 'rb') do |f|
    f.seek(start)
    f.read(size) || ""
  end
rescue => e
  STDERR.puts "Failed to read file: #{e.message}"
  exit 1
end

def trim_truncated_lines(data, start)
  lines = data.lines
  lines.shift if start > 0 && lines.any?
  lines.pop unless lines.empty? || lines.last.end_with?("\n")
  lines.join
end

begin
  data = read_slice(options[:file], options[:start], options[:size])
  data = trim_truncated_lines(data, options[:start]) if options[:full_lines_only]
  STDOUT.write(data)
rescue => e
  STDERR.puts "Unexpected error: #{e.message}"
  exit 1
end
